dandi/download.py

```diff
@@ -878,33 +878,40 @@ def _download_zarr(
     # Avoid heavy import by importing within function:
     from .support.digests import get_zarr_checksum
 
-    download_gens = {}
-    entries = list(asset.iterfiles())
+    # we will collect them all while starting the download
+    # with the first page of entries received from the server.
+    entries = []
     digests = {}
+    pc = ProgressCombiner(zarr_size=asset.size)
 
     def digest_callback(path: str, algoname: str, d: str) -> None:
         if algoname == "md5":
             digests[path] = d
 
-    for entry in entries:
-        etag = entry.digest
-        assert etag.algorithm is DigestType.md5
-        download_gens[str(entry)] = _download_file(
-            entry.get_download_file_iter(),
-            download_path / str(entry),
-            toplevel_path=toplevel_path,
-            size=entry.size,
-            mtime=entry.modified,
-            existing=existing,
-            digests={"md5": etag.value},
-            lock=lock,
-            digest_callback=partial(digest_callback, str(entry)),
-        )
+    def downloads_gen():
+        for entry in asset.iterfiles():
+            entries.append(entry)
+            etag = entry.digest
+            assert etag.algorithm is DigestType.md5
+            yield pairing(
+                str(entry),
+                _download_file(
+                    entry.get_download_file_iter(),
+                    download_path / str(entry),
+                    toplevel_path=toplevel_path,
+                    size=entry.size,
+                    mtime=entry.modified,
+                    existing=existing,
+                    digests={"md5": etag.value},
+                    lock=lock,
+                    digest_callback=partial(digest_callback, str(entry)),
+                ),
+            )
+        pc.file_qty = len(entries)
 
-    pc = ProgressCombiner(zarr_size=asset.size, file_qty=len(download_gens))
     final_out: dict | None = None
     with interleave(
-        [pairing(p, gen) for p, gen in download_gens.items()],
+        downloads_gen(),
```

Comment:
`interleave()` evaluates/consumes the iterable in its entirety before returning, so you haven't accomplished anything by making `downloads_gen()` into a generator.

In order to add downloads to `interleave()` as they're yielded by `asset.iterfiles()`, you'll need to:

* Replace the call to `interleave()` with a directly-constructed `Interleaver` instance
* In separate thread, call the `Interleaver`'s `submit()` method whenever a new file is yielded by `asset.iterfiles()`, and then call its `finalize()` method at end of iteration.

[See `interleave`'s docs for more information.](https://github.com/jwodder/interleave?tab=readme-ov-file#api)

Reply:
would you be so kind to approach this or otherwise address this issue the way you like it while also ideally reducing that delay / not demanding full listing to arrive to start the download?

Reply:
or may be we could merge this (as resolves the issue) and then improve on top of it?

Reply:
Getting the error handling right for the solution I suggested would be very difficult, so we should probably just merge the current version for now.

Reply:
ok, let's proceed!  Do you want me to create a dedicated issue or you keep this on your plate?

Reply:
Create a dedicated issue.